### ISSUE TITLE
ICE librustc_const_eval/_match.rs:261: bad constructor […] for adt NonZeroU32

### ISSUE THREAD

**Issue by SimonSapin** [rust-lang/rust#48279]
Test case:

```rust
#![feature(const_fn)]

#[derive(PartialEq, Eq)]
pub struct NonZeroU32 {
    value: u32
}

impl NonZeroU32 {
    const unsafe fn new_unchecked(value: u32) -> Self {
        NonZeroU32 { value }
    }
}

//pub const FOO_ATOM: NonZeroU32 = unsafe { NonZeroU32::new_unchecked(7) };
pub const FOO_ATOM: NonZeroU32 = unsafe { NonZeroU32 { value: 7 } };

fn main() {
    match None {
        Some(FOO_ATOM) => {}
        _ => {}
    }
}
```

Backtrace below.

Using `NonZeroU32 { value: 7 }` without `unsafe {…}` compiles (but of course requires the private field to be visible, unlike `new_unchecked`).

```rust
error: internal compiler error: librustc_const_eval/_match.rs:261: bad constructor ConstantValue(Const { ty: NonZeroU32, val: Aggregate(Struct([(value, Const { ty: u32, val: Integral(U32(7)) })])) }) for adt NonZeroU32

thread 'rustc' panicked at 'Box<Any>', librustc_errors/lib.rs:535:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at libstd/sys_common/backtrace.rs:59
             at libstd/panicking.rs:380
   3: std::panicking::default_hook
             at libstd/panicking.rs:396
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:576
   5: std::panicking::begin_panic
   6: rustc_errors::Handler::bug
   7: rustc::session::opt_span_bug_fmt::{{closure}}
   8: rustc::ty::context::tls::with_opt::{{closure}}
   9: <std::thread::local::LocalKey<T>>::try_with
  10: <std::thread::local::LocalKey<T>>::with
  11: rustc::ty::context::tls::with
  12: rustc::ty::context::tls::with_opt
  13: rustc::session::opt_span_bug_fmt
  14: rustc::session::bug_fmt
  15: rustc_const_eval::_match::Constructor::variant_index_for_adt
  16: rustc_const_eval::_match::constructor_sub_pattern_tys
  17: rustc_const_eval::_match::is_useful_specialized
  18: <core::iter::Map<I, F> as core::iter::iterator::Iterator>::try_fold::{{closure}}
  19: rustc_const_eval::_match::is_useful
  20: rustc_const_eval::_match::is_useful_specialized
  21: <core::iter::Map<I, F> as core::iter::iterator::Iterator>::try_fold::{{closure}}
  22: rustc_const_eval::_match::is_useful
  23: <rustc_const_eval::check_match::MatchVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr
  24: <rustc_const_eval::check_match::MatchVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr
  25: rustc_const_eval::check_match::check_match
  26: rustc::dep_graph::graph::DepGraph::with_task_impl
  27: rustc::ty::maps::<impl rustc::ty::maps::queries::check_match<'tcx>>::force
  28: rustc::ty::maps::<impl rustc::ty::maps::queries::check_match<'tcx>>::try_get
  29: rustc::ty::maps::TyCtxtAt::check_match
  30: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::check_match
  31: rustc_const_eval::check_match::check_crate
  32: rustc::ty::context::TyCtxt::create_and_enter
  33: rustc_driver::driver::compile_input
  34: rustc_driver::run_compiler

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.25.0-nightly (3ec5a99aa 2018-02-14) running on x86_64-unknown-linux-gnu
```

**Comment by oli-obk** [comment]
Should be fixed by #46882 (I added a regression test)

**Comment by SimonSapin** [comment]
This is fixed in `master` (presumably by #46882). Thanks!
